agent/lib/src/external_api/service.rs

```diff
@@ -0,0 +1,264 @@
+                    }
+                }
+                Err(ref e) if e.kind() == std::io::ErrorKind::TimedOut => (),
+                Err(e) => error!("{:?}", e),
+            }
+        }
+
+        bytes_read = 0;
+
+        while bytes_read < data_size {
+            match self.serial_port.read(&mut buf) {
+                Ok(t) => {
+                    if t > 0 {
+                        bytes_read += t;
+                        data_received.extend_from_slice(&buf[..t]);
+                        debug!("Received {} bytes", t);
+                    }
+                }
+                Err(ref e) if e.kind() == std::io::ErrorKind::TimedOut => (),
+                Err(e) => error!("{:?}", e),
+            }
+        }
+
+        debug!("Final received data: {:?}", data_received);
+        debug!("Total bytes read {:?}", bytes_read);
+
+        let code_entry = self.parse_json_payload(&data_received)?;
+
+        info!("Code entry: {:?}", code_entry);
+
+        // Flush the serial port
+        self.serial_port
+            .flush()
+            .map_err(|e| anyhow!("Failed to flush serial port: {}", e))?;
+
+        Ok(code_entry)
+    }
+
+    pub fn parse_json_payload(&mut self, data: &[u8]) -> Result<RequestMessage> {
+        // Convert the data vector to a codeEntry struct
+        let request_message: RequestMessage = serde_json::from_slice(data)
+            .map_err(|e| anyhow!("Failed to parse JSON payload: {}", e))?;
+
+        info!("Code request message: {:?}", request_message);
+
+        Ok(request_message)
+    }
+
+    pub fn send_status_message(&mut self) -> Result<()> {
+        let status_message: StatusMessage = StatusMessage::new();
+        let status_message_json = serde_json::to_string(&status_message)
+            .map_err(|e| anyhow!("Failed to serialize status message: {}", e))?;
+        self.write_to_serial(&status_message_json)?;
+        Ok(())
+    }
+
+    pub fn send_response_message(&mut self, response_message: ResponseMessage) -> Result<()> {
+        let code_json = serde_json::to_string(&response_message)
+            .map_err(|e| anyhow!("Failed to stringify response message : {}", e))?;
+
+        // Write the JSON to the serial port
+        self.write_to_serial(&code_json)?;
+
+        info!(
+            "Response message written to serial port: {:?}",
+            response_message
+        );
+        Ok(())
+    }
+
+    pub fn write_to_serial(&mut self, data: &str) -> Result<()> {
+        info!("Writing to serial port: {}", self.serial_path);
+
+        // Convert the string to a byte array
+        let message = Message::new(data.to_string()).to_bytes();
+        let buf = message.as_slice();
+
+        // Write the byte array to the serial port
+        self.serial_port
+            .write_all(buf)
+            .map_err(|e| anyhow!("Failed to write to serial port: {}", e))?;
+
+        // In order to still be readable by ``readline`` on the other side, we add a carriage return
```

Comment:
If the `other side` decides to use `readline` to read the agent's output, it's none of the agent's business. You should not add unspecified data to your protocol.

Reply:
Normally I would absolutely agree with you.
However since the API uses [BufReader::read_line](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.read_line) they need a carriage return, else the API just blocks meaning that we have to send it. 

I know this looks ridicoulous, I tried to tell them to stick to what we have defined.